lib/prism/parse_result.rb

```diff
@@ -106,6 +114,39 @@ def find_line(byte_offset)
     end
   end
 
+  # Specialized version of Prism::Source for source code that includes ASCII
+  # characters only. This class is used to apply performance optimizations that
+  # cannot be applied to sources that include multibyte characters. Sources that
+  # include multibyte characters are represented by the Prism::Source class.
+  class ASCIISource < Source
+    # Return the character offset for the given byte offset.
+    def character_offset(byte_offset)
+      byte_offset
+    end
```

Comment:
I think we could use this logic directly in `Source` like `return byte_offset if @source.ascii_only?` (the result of `ascii_only?` is cached via String code range) or `return byte_offset if @ascii_only`.
Then there wouldn't be the concern that `Source.new` might be used instead of `Source.for` and that would then be needlessly slower.

OTOH, maybe users aren't supposed to use `Source.new`/`Source.for` directly anyway? (since e.g. `Prism.parse` takes a String and not a Source)

Reply:
Yeah `Source` would be package-private if that were allowed in Ruby. I wanted us to specialize on creation of the object because the `source` field is also not meant to change.